MWFTP/clip.py

```diff
@@ -0,0 +1,46 @@
+"""Provides functionality for clipping extreme values.
+
+Based on NormDistCapping.fit_transform().
+"""
+import pandas as pd
+
+
+def clip(data: pd.Series, window_frac: float, stds_clip: float,
+         stds_remove: float) -> pd.DataFrame:
+    """Returns a copy of the series with some of its values possibly removed or
+    clipped, according to their distance from the mean.
+
+    Values stds_remove or more away from the mean are removed. Values stds_clip or
+    more away from mean are set to (mean +- stds_clip)."""
+
+    assert 0 < window_frac <= 1, f'{window_frac}'
+
+    # Find the densest window in the series (lowest span).
+    # We consider only the values within this window for calculating mean and std,
+    # to avoid extreme values.
+    series = data.sort_values()
+    series = series[~series.isna()]
+    window_size = round(len(series) * window_frac)
+    windows = series.rolling(window_size)
+    spans = windows.max() - windows.min()
+    spans = spans[window_size - 1:]  # Remove prefix of NAs.
```

Comment:
what would happen if there is no na?

Reply:
`rolling()` functions pad their output with NAs, maybe to make its length equal to the input's. There are n-w+1 windows, so there are w-1 NAs by definition. Here I remove them to retain only the real results.